scene/animation/animation_tree.cpp

```diff
@@ -2244,6 +2321,8 @@ void AnimationTree::_bind_methods() {
 
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "active"), "set_active", "is_active");
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "process_callback", PROPERTY_HINT_ENUM, "Physics,Idle,Manual"), "set_process_callback", "get_process_callback");
+	ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "filter_presets", PROPERTY_HINT_TYPE_STRING, vformat("%d/%d:%s", Variant::OBJECT, PROPERTY_HINT_RESOURCE_TYPE, "AnimationTrackFilter")), "set_filter_presets", "get_filter_presets");
+
```

Comment:
Preset is primarily used only for Editor, so it should not be stored as a property in BlendTree. At worst, it must be enabled only in the TOOLS environment (with #ifdef TOOLS_ENABLED, see `Animation::HandleMode` FYI). Or it needs to be changed so that it can be set for `AnimationBlendTreeEditor` instead of AnimationTree's Inspector.

Reply:
Or Probably we may not need the Preset feature in the first place, it can be solved by just load the Resourced filter and make it unique.

Reply:
The reason why I let presets as a property of `AnimationTree` as follow:
1. A set of presets are only meaningful for a series of animations which are included in a `AnimationPlayer`, but only used by `AnimationNode`s in  `AnimationTree`.
2. Typically, a set of presets is work with a character, it may not be suitable for another charcter, and a `AnimationTree` is work wich a charcter, too.

So, I think it should be store which `AnimationTree`.
And for flexibility reasons, I support load a Filer resource in `AnimationBlendTreeEditor`.

-------------------------------------------------------------

Currently, I agree that presets as property should only work in editor.

But I have a proposal, can we support a replaceable `AnimationNode`? Like `Animation Layer` or `Animation Layer Interface` in unreal engine.
We can replace a `AnimationNode` in code, typically, we want to replace a partial `AnimationNodeBlendTree` which describe a general logic of animation.
If yes, to adapt to the target `AnimationTree`, children nodes of replacable `AnimationNode` should change their filter, and developers can get filter presets from `AnimationTree` in code.

Reply:
If you want to replace filters in multiple AnimationNodes per character, you can specify references to the same rewritable filter resource in multiple AnimationNodes and copy the settings of other filters to that filter resource. And if you are concerned about the performance of copying an array, you can replace them by nesting the resources for the filters.

For example:
- Specify a BlendSetting resource for the AnimationNode
- BlendSetting resource has a FilterSetting resource
- FilterSetting resource has a Vector<StingName>

Reply:
I think this is not reasonable, the `AnimationNode` which will be replaced is not concerned about it will be used in which `AnimationTree` or character, which filter should be used can't be specified in editor.

Reply:
Yes, it means that an editor must be implemented to set the resource to the AnimationNode. However, implementing presets in the AnimationTree is not the right solution.

Reply:
The core issue is how developers can find a filter preset resource.
The first way is save filter resources in file system and manage by developers themselves, and load preset from file system.
The second way is add a presets property for `AnimationTree`, developers not need to manage presets by using this way, and can select preset of specific character directly in `AnimationBlendTreeEditor`, because `AnimationBlendTreeEditor` is specified the editing `AnimationTree`.

I think we need more people to participate in this discussion, to decide whether the second way is necessary.

Reply:
In my opinion, if you want to reuse the AnimationTree with multiple characters and apply a different filter to each one, the correct way is to use `AnimationNode::get_parameter_list()` to implement the filter as a parameter that can be set for each Node individually, like blend amount. Then, the filter should be set in the "parameters" section of the AnimationTree inspector.

Reply:
It sound good, but there have some questions confuse me:
1. The process of get parameter need multiple check and hash, I don't know low level mechanism of this process, will it  significantly reduce performance?
2. How can we setup filters for `AnimationNode`s which can be replaced in editor? The structure/paths of it's child nodes can't be unable to be determined in design time.
Or we should change the "parameters"' mechanism of `AnimationTree`, make it independent on path, it will be a huge change.

Reply:
![image](https://github.com/godotengine/godot/assets/61624558/6d191ff5-9d18-4711-b439-6d8f2731bdcc)

In the image, the structure of "SpecalBlendTree 1" and "SpecalBlendTree 2" is invisable in the `AnimationTree`, and can replace "PlaceHolderBlendTree" at run time, it seem kind of like the concept of "Interface" in programing, I have not idea to describe this use case clearly.

Reply:
As mentioned above, a BlendTree has its direct children as a list, so the parameters in it should be visible in the parameters section of the AnimationTree inspector, even if they are not connected.

If there are many unconnected BlendTrees, this means that the number of parameters will also be large. However, this is also applied to blend amounts, etc., so this has nothing to do with making the filter a parameter.

Reply:
No, two BlendTree at left is not in this graph, I just place them in one Image to tell you that there have two node resource can be used to replace, in my imagine, they should be loaded from file system.

Reply:
Parameters that each AnimationTree has, such as blend amount, are not stored in the BlendTree resource.

When a BlendTree is reconnected, the AnimationTree will rebuild the hashmap. You will then need to set the parameters.

You probably set the blend amount every frame, so you may not care, but those values need to be set again after the hashmap is rebuilt.

If we make the filter a parameter, we need to set the filter after rebuilding the hashmap in the same way. That is the role of gdscript and it should not be difficult if filter becomes a resource.

Reply:
If we can't implement a mechanism that replacing an `AnimationNode` like `interface` and setup its parameters ( such as filter) in editor, let "filter" to work like "blend_amount" maybe meaningless, the way of editing filter is fine currently, developer can check the relationship between the Node and its inputs, and avoid increasing complexity of `AnimationTree`‘s parameters section.

Any way, "replaceable AnimationNode" or "AnimationNode Interface" is another topic, it is just an unrealized concept that still not be implemented (yes, we can use a serial of operations to replace a `AnimationNode`, but not an official implement). We should not be stuck here.

Let's return the core issue, should we clearfy the concept that "a set of filter presets is belong to an `AnimationTree`", or prefer to "filter just a resource, you can use it for any `AnimationNode`, any `AnimationTree` or any character".

Reply:
Personally, I strongly disagree with the preset as it does not conform to the current AnimationTree design and looks too hacky. If the latter, I can personally approve.

Reply:
OK, I remove the `preset` concept, and implement an inspector plugin to edit tracks of `AnimaionTrackFilter` if there has selected an `AnimationTree`.

Reply:
I have also discussed this with @lyuma and we agreed to make the filter a Resource.

So now it's just a matter of how to implement it, but if we store it in a parameter, we may not need `AnimationNode::has_filter/set_filter`. Instead, we need to somehow show up a button for editing in AnimationBlendTreeEdit, but if that is difficult, you may want to continue to use `has_filter` and try it once and refactor it later.

Reply:
Your conclusion is that implement filter as a parameter is better, right?

I think It is not too hard, but I still have some concerns about performance, animation system is one of significant consumption in 3D game, we should try our best to implement it as static style.

Reply:
Yeah, I think it would make more sense to make filter a parameter in the case where you want to share the AnimationTree with several characters and apply different filters. But well, it would be possible to make it a different PR.

Reply:
Oh yes, I see what you mean.
I will change my code later.

Reply:
To make it easier for developers to extend `AnimationNode`, I think  `AnimationNode::has_filter()` should be retained.

Reply:
One idea is to add a class like AnimationNodeFilter as BaseClass that extends AnimationNodeSync with the assumption that it has filters. After returning the filter parameters, use a super function or similar to return the individual parameters.